### PR TITLE
Allow ignoring soft wraps when moving to line ends

### LINKED ISSUE #10888: Keyboard end does not go at the end but at first line end of wrap
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

image

As shown here
It should go to the end of the wrap when I hit <kbd>Fin (End)</kbd> on my keyboard.
Vscode works fine with this behavior. I guess this is an implementation bug.

```json

  "soft_wrap": "preferred_line_length",
  "preferred_line_length": 180,
```

### Environment

Zed: v1.0.0 (Zed Dev a0fa8a489bb4af98059e5a064c7fac2a77b49aff)
OS: Debian 12
Memory: 62.4 GiB
Architecture: x86_64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by tverghis**
Release Notes:

- Fixed #10888

This patch addresses behavior of `Editor::move_to_{beginning|end}_of_line`. It adds a setting, `stop_at_soft_wraps` when defining a keymap for the `editor::MoveToBeginningOfLine` and `editor::MoveToEndOfLine` actions. When `true`, it causes movement to the either end of the line (via, for example Home or End), to go to the logical end, as opposed to the nearest soft wrap point in the respective direction.

### CODE REVIEW

**Comment by tverghis** [comment]
@cla-bot check

**Comment by tverghis** [comment]
Looks like this broke some assertions in `editor_tests::test_beginning_end_of_line`. I'll try to fix that, and add some additional tests for when `stop_at_soft_wraps` is true (old behavior).

File: crates/editor/src/actions.rs [hunk 1]
```diff
@@ -106,10 +118,12 @@ impl_actions!(
         SelectNext,
         SelectPrevious,
         SelectToBeginningOfLine,
+        MoveToBeginningOfLine,
```
Comment (SomeoneToIgnore) [inline crates/editor/src/actions.rs:121]: Let's sort these alphabetically.

File: crates/editor/src/actions.rs [hunk 2]
```diff
@@ -13,6 +13,12 @@ pub struct SelectPrevious {
     pub replace_newest: bool,
 }
 
+#[derive(PartialEq, Clone, Deserialize, Default)]
+pub struct MoveToBeginningOfLine {
+    #[serde(default)]
```
Comment (SomeoneToIgnore) [inline crates/editor/src/actions.rs:18]: This changes the current default, which corresponded to `stop_at_soft_wraps: true`, let's keep it unchanged instead and use https://github.com/zed-industries/zed/blob/8a79535b84dfee2f88a3ad865465893ec3cdf540/crates/language/src/language_settings.rs#L421 or similar idea for that.

**Review by SomeoneToIgnore — COMMENTED** [review]
Looks nice, let's add a small test, similar to what `test_beginning_end_of_line` has (or just another few lines straight there)?

File: crates/editor/src/actions.rs (hunk 2 above)
Comment (tverghis) [reply crates/editor/src/actions.rs:18]: Is the current behavior (i.e., stopping at soft-wraps) desired, or is it a bug? Defaulting to `false` would be a change to existing behavior, but maybe that is fine if the behavior is unintended?

If the current behavior is intended, then this PR becomes more of an enhancement ("allow configuring stop-at-soft-wrap when moving to line ends") than a bug fix. In that case I should update the PR title and release notes. I don't mind either way -- let me know what you think!

Comment (SomeoneToIgnore) [reply crates/editor/src/actions.rs:18]: Yeah, I'd say this looks more like an improvement and the old behavior is intended (at least, for some group of people).

File: crates/editor/src/actions.rs (hunk 1 above)
Comment (tverghis) [reply crates/editor/src/actions.rs]: Done

**Comment by tverghis** [comment]
@SomeoneToIgnore I believe my PR is ready for re-review now. Thanks for taking a look!

**Review by SomeoneToIgnore — COMMENTED** [review]
Thank you, the changes look complete now — let's fix the tests and have it merged.

**Comment by tverghis** [comment]
Sorry for the delay on this.

Unfortunately, it seems that some tests have different results based on whether it runs on CI or on my machine. The tests that are affected by my changes pass on my machine. FWIW, I am on Ubuntu.
```bash
failures:
    display_map::block_map::tests::test_blocks_on_wrapped_lines
    display_map::tests::test_chunks_with_soft_wrapping
    editor_tests::test_add_selection_above_below
    editor_tests::test_beginning_end_of_line
    editor_tests::test_move_cursor_multibyte
    editor_tests::test_prev_next_word_bounds_with_soft_wrap
    hover_links::tests::test_hover_type_links

test result: FAILED. 183 passed; 7 failed; 0 ignored; 0 measured; 0 filtered out; finished in 149.87s
```

It _appears_ that many of the failures likely have something to do with wrapping. Is text wrapping dependent on the machine or OS it runs on?

Anyway, to fix the problem, I will update the expected values to try and make CI happy.

**Comment by tverghis** [comment]
@SomeoneToIgnore tests have been updated

**Comment by SomeoneToIgnore** [comment]
Oh, we do not run tests on Ubuntu yet and I've just stumbled onto similar issues yesterday.
Interesting to see how few tests are left to fix, and sorry for the bad experience on your side — our Linux support is still in the experimental phase.

Thank you for altering those nonetheless, I think it's fine to use mac values for now and fix the Linux tests later.

**Comment by SomeoneToIgnore** [comment]
There seems to be another failure, I'll look into this and fix it today.
Hopefully you've ticked the "allow edits from maintainers" in your PR, or you'll have to apply a patch from me 🙂

**Comment by SomeoneToIgnore** [comment]
So, checked `map.max_point()` in the `fn line_end` and effectively failed the test on CI, and it returns `DisplayPoint(2, 5)` whilst on your system, that was `DisplayPoint(2, 6)` presumably.

Given the particular set of tests failing, I wonder if that's an off-by-one or other bug in the way we treat display width somewhere in the depth.
We initialize the tests with the same screen width on both platforms, so it's not the test infra code.
I will try to look at it, as interested in fixing Linux tests soon, but no promises yet.

The end diff is just
```
diff --git a/crates/editor/src/editor_tests.rs b/crates/editor/src/editor_tests.rs
index de2161b91..17f2a3c36 100644
--- a/crates/editor/src/editor_tests.rs
+++ b/crates/editor/src/editor_tests.rs
@@ -1288,14 +1288,14 @@ fn test_beginning_end_of_line_ignore_soft_wrap(cx: &mut TestAppContext) {
         // next display line.
         view.move_to_end_of_line(&move_to_end, cx);
         assert_eq!(
-            vec![DisplayPoint::new(2, 6)..DisplayPoint::new(2, 6),],
+            vec![DisplayPoint::new(2, 5)..DisplayPoint::new(2, 5),],
             view.selections.display_ranges(cx)
         );
 
         // Moving to the end of the line again should be a no-op.
         view.move_to_end_of_line(&move_to_end, cx);
         assert_eq!(
-            vec![DisplayPoint::new(2, 6)..DisplayPoint::new(2, 6),],
+            vec![DisplayPoint::new(2, 5)..DisplayPoint::new(2, 5),],
             view.selections.display_ranges(cx)
         );
     });
```

**Comment by tverghis** [comment]
@SomeoneToIgnore thanks so much for patching this up and getting it merged!

**Pull request merged**